torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -367,6 +367,15 @@ Val* castOp(DataType dtype, Val* v1) {
     return set(v1);
   }
 
+  // There are no direct cast for some combinations of dtypes
+  // for example integers to fp16, bfloat16 and back
+  // so we need to try doing a conversion to float first.
+  if (dtype != DataType::Float &&
+      cast_func_str(std::make_pair(v1->getDataType().value(), dtype)) ==
```

Comment:
This feels a little bit not ideal. Lol.

Should we actually add spot support for cast functions instead? cc'ing @kevinstephano

Reply:
I was hesitant to modify `cast_func_str` as it appears also in codegen.cpp, ir_iostream.cpp. One hop through the float dtype seemed like a clear choice to me.

I can also move the changes to `cast_func_str` if that's preferred.

Reply:
You'll need to update `supported_casts2string` instead and also plumb some inline ptx to the runtime files for conversion.

Totally agree that the changes in this PR is more general and cleaner. I'm just not sure if we DO want this hammer :laughing: Already approved, so feel free to merge as-is. I'm just seeking some second opinion from Kevin, who has much better instincts than I do on how cuda code translates to perf.

Reply:
@kevinstephano please see ^^ and comment

Reply:
We talked about this last week where we were interested in looking for a PTX version that does this officially.  The model where this happens is a little weird in the first place as it looks like some generic editing for autocast in torchbench was done that might have not been totally desirable but it isn't crazy that a real user might do this.